core/src/main/java/kafka/server/share/SharePartitionManager.java

```diff
@@ -0,0 +1,418 @@
+                log.debug("Share session context with key {} isSubsequent {} returning {}", responseShareSessionKey,
+                        false, partitionsToLogString(shareFetchDataWithMaxBytes.keySet()));
+
+                context = new ShareSessionContext(reqMetadata, shareFetchDataWithMaxBytes);
+                log.debug("Created a new ShareSessionContext with {}. A new share session will be started.",
+                        partitionsToLogString(shareFetchDataWithMaxBytes.keySet()));
+            }
+        } else {
+            synchronized (cache) {
+                ShareSessionKey key = shareSessionKey(groupId, reqMetadata.memberId());
+                ShareSession shareSession = cache.get(key);
+                if (shareSession == null) {
+                    log.debug("Share session error for {}: no such share session found", key);
+                    throw Errors.SHARE_SESSION_NOT_FOUND.exception();
+                } else {
+                    if (shareSession.epoch != reqMetadata.epoch()) {
+                        log.debug("Share session error for {}: expected epoch {}, but got {} instead", key,
+                                shareSession.epoch, reqMetadata.epoch());
+                        throw Errors.INVALID_SHARE_SESSION_EPOCH.exception();
+                    } else {
+                        Map<ShareSession.ModifiedTopicIdPartitionType, List<TopicIdPartition>> modifiedTopicIdPartitions = shareSession.update(
+                                shareFetchDataWithMaxBytes, toForget);
+                        cache.touch(shareSession, time.milliseconds());
+                        shareSession.epoch = ShareFetchMetadata.nextEpoch(shareSession.epoch);
+                        log.debug("Created a new ShareSessionContext for session key {}, epoch {}: " +
+                                        "added {}, updated {}, removed {}", shareSession.key(), shareSession.epoch,
+                                partitionsToLogString(modifiedTopicIdPartitions.get(
+                                        ShareSession.ModifiedTopicIdPartitionType.ADDED)),
+                                partitionsToLogString(modifiedTopicIdPartitions.get(ShareSession.ModifiedTopicIdPartitionType.UPDATED)),
+                                partitionsToLogString(modifiedTopicIdPartitions.get(ShareSession.ModifiedTopicIdPartitionType.REMOVED))
+                        );
+                        context = new ShareSessionContext(reqMetadata, shareSession);
+                    }
+                }
+            }
+        }
+        return context;
+    }
+
+    private ShareSessionKey shareSessionKey(String groupId, Uuid memberId) {
+        return new ShareSessionKey(groupId, memberId);
+    }
+
+    String partitionsToLogString(Collection<TopicIdPartition> partitions) {
+        return FetchSession.partitionsToLogString(partitions, log.isTraceEnabled());
```

Comment:
Also we do have a same method defined in `ShareFetchContext`. Do we need both?

Reply:
functionally they are the same, its just that the output string is different when `isTraceEnabled()` is true or false for different logger objects. So, it gives us that flexibility.